### PR TITLE
Fix LedgerDB OnDisk tests

### PR DESCRIPTION
**Pull request by dnadales**
After the fix introduced in #3827 the LedgerDB OnDisk tests started failing because the tables in the backing store did not contain the values of the initial ledger state.

In addition, we now model the flushing aspect. The model did not take into account the flushing aspect. As a result, the model and SUT will differ in the slot at which snapshots were taken, or the model would take a snapshot whereas the SUT did not.
    
This patch introduces a `mockLatestFlushedPoint` field in the `Mock` that keeps track of the point at which a immutable tip in the sequence of ledger states was flushed to disk. The runMock function was updated to account for this new field.

# Checklist

- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

**Review by jorisdral — COMMENTED** [review]
LGTM :) I mostly posed questions to clarify some of the comments you wrote

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs [hunk 1]
```diff
@@ -706,13 +743,42 @@ runMock cmd initMock =
...
+    go Restore       mock = (Restored (initLog, forgetLedgerTables $ cur (mockLedger mock'')), mock'')
+      -- When we restore to the most recent valid snapshot, the point at which a
+      -- immutable tip was flushed to disk becomes the tip of the restoraton
+      -- point, since any flushes that occured after this point are now invalid
+      -- because we restored a snapshot.
+      --
+      -- TODO: This depends on the SUT not flushing during restoration. At the
+      -- moment we flush every 100 blocks, and the chains we generate are far
+      -- below that (about less than 40 elements). At the moment we do not model
+      -- the flushing during restoration aspect here.
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs:755]: The Model wouldn't be able to keep track of the flushes that happen in `initStartingWith` as `Restore` is only one command from the point of view of the test. We would have to factor out that value into the config and use that config to make the Model know that if it replays more than X blocks, it should also account for a flush.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 2]
```diff
@@ -50,7 +51,7 @@ module Ouroboros.Consensus.Storage.LedgerDB.InMemory (
   , ledgerDbChangelog
   , ledgerDbCurrent
   , ledgerDbFlush
-  , ledgerDbOldest
+  , ledgerDbLastFlushState
```
Comment (Jasagredo) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:54]:
```suggestion
  , ledgerDbLastFlushedState
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs [hunk 3]
```diff
@@ -659,17 +661,18 @@ ledgerDbAnchor =
...
 -- before calling this function
-ledgerDbOldest :: forall l.
+ledgerDbLastFlushState :: forall l.
      ( StandardHash (l EmptyMK)
      , GetTip (l EmptyMK)
      , StowableLedgerTables l
      )
   => LedgerDB l -> l EmptyMK
-ledgerDbOldest db =
+ledgerDbLastFlushState db =
```
Comment (Jasagredo) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs:670]:
```suggestion
ledgerDbLastFlushedState db =
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs [hunk 4]
```diff
@@ -983,12 +987,12 @@ takeSnapshot tracer hasFS backingStore encLedger db =
...
         else do
-          writeSnapshot hasFS backingStore encLedger snapshot oldest
+          writeSnapshot hasFS backingStore encLedger snapshot lastFlushed
           traceWith tracer $ TookSnapshot snapshot tip
           return $ Just (snapshot, tip)
   where
-    oldest :: ExtLedgerState blk EmptyMK
-    oldest = ledgerDbOldest db
+    lastFlushed :: ExtLedgerState blk EmptyMK
+    lastFlushed = ledgerDbLastFlushState db
```
Comment (Jasagredo) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/OnDisk.hs:995]:
```suggestion
    lastFlushed = ledgerDbLastFlushedState db
```

**Review by Jasagredo — APPROVED** [review]
Approved with some comments

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs [hunk 5]
```diff
@@ -706,13 +743,42 @@ runMock cmd initMock =
     go Current       mock = (Ledger (forgetLedgerTables $ cur (mockLedger mock)), mock)
     go (Push b)      mock = first MaybeErr $ mockUpdateLedger (push b)      mock
     go (Switch n bs) mock = first MaybeErr $ mockUpdateLedger (switch n bs) mock
-    go Restore       mock = (Restored (initLog, forgetLedgerTables $ cur (mockLedger mock')), mock')
+    go Restore       mock = (Restored (initLog, forgetLedgerTables $ cur (mockLedger mock'')), mock'')
+      -- When we restore to the most recent valid snapshot, the point at which a
+      -- immutable tip was flushed to disk becomes the tip of the restoraton
+      -- point, since any flushes that occured after this point are now invalid
+      -- because we restored a snapshot.
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/OnDisk.hs:750]: No, this does not refer to snapshots :)

**Pull request merged**